client/app/visualizations/EditVisualizationDialog.jsx

```diff
@@ -0,0 +1,211 @@
+
+import { toastr } from '@/services/ng';
+import recordEvent from '@/services/recordEvent';
+
+// ANGULAR_REMOVE_ME Remove when all visualizations will be migrated to React
+import { cleanAngularProps } from '@/lib/utils';
+
+class EditVisualizationDialog extends React.Component {
+  static propTypes = {
+    dialog: DialogPropType.isRequired,
+    query: PropTypes.object.isRequired, // eslint-disable-line react/forbid-prop-types
+    visualization: PropTypes.object, // eslint-disable-line react/forbid-prop-types
+    queryResult: PropTypes.object.isRequired, // eslint-disable-line react/forbid-prop-types
+  };
+
+  static defaultProps = {
+    visualization: null,
+  };
+
+  constructor(props) {
+    super(props);
+
+    const { visualization, queryResult } = this.props;
+
+    const isNew = !visualization;
+
+    const config = isNew ? getDefaultVisualization()
+      : registeredVisualizations[visualization.type];
+
+    // it's safe to use queryResult right here because while query is running -
+    // all UI to access this dialog is hidden/disabled
+    const data = {
+      columns: queryResult.getColumns(),
+      rows: queryResult.getData(),
+    };
+
+    const options = config.getOptions(isNew ? {} : visualization.options, data);
+
+    this.state = {
+      isNew, // eslint-disable-line
+      data,
+
+      type: config.type,
+      canChangeType: isNew, // cannot change type when editing existing visualization
+      name: isNew ? config.name : visualization.name,
+      nameChanged: false,
+      originalOptions: cloneDeep(options),
+      options,
+
+      saveInProgress: false,
+    };
+  }
+
+  setVisualizationType(type) {
+    this.setState(({ isNew, name, nameChanged, data }) => {
+      const { visualization } = this.props;
+      const config = registeredVisualizations[type];
+      const options = config.getOptions(isNew ? {} : visualization.options, data);
+      return {
+        type,
+        name: nameChanged ? name : config.name,
+        options,
+      };
+    });
+  }
+
+  setVisualizationName(name) {
+    this.setState({
+      name,
+      nameChanged: true,
+    });
+  }
+
+  setVisualizationOptions = (options) => {
+    this.setState({ options: extend({}, options) });
+  };
+
+  dismiss() {
+    const { nameChanged, options, originalOptions } = this.state;
+
+    const optionsChanged = !isEqual(cleanAngularProps(options), originalOptions);
+    if (nameChanged || optionsChanged) {
+      Modal.confirm({
+        title: 'Visualization Editor',
+        content: 'Are you sure you want to close the editor without saving?',
+        okText: 'Yes',
+        cancelText: 'No',
+        onOk: () => {
+          this.props.dialog.dismiss();
+        },
+      });
+    } else {
+      this.props.dialog.dismiss();
+    }
+  }
+
+  save() {
+    const { query } = this.props;
+    const { type, name, options } = this.state;
+
+    const visualization = {
+      ...extend({}, this.props.visualization),
```

Comment:
This any different from `...this.props.visualization`?

Reply:
@ranbena In fact - yes, because `this.props.visualization` may be `null`, but I have no idea why I did this instead of just 
```javascript
const visualization = extend({}, this.props.visualization, newVisualization(type));
```
:sweat_smile:

Reply:
And BTW there is a bug - it should be
```javascript
const visualization = extend({}, newVisualization(type), this.props.visualization);
```
:thinking: